### PR TITLE
#166789996 Users Should Be Able To Create Their Profiles

### PR DESCRIPTION
**Pull request by IsaiahRn**
#### What does this PR do?

- Add a feature to create a user profile

#### Description of Task to be completed

- After successful registration, a user profile must be created.

- User should be able to view and Edit/Update his own profile.

#### How should this be manually tested?

1. Run the following commands in your terminal

```
git clone https://github.com/andela/ah-92explorers-backend.git
cd ah-92explorers-backend
git checkout ft-create-profiles-166789996
npm install
npm start
```

2. Go to Postman

- Register a user via `localhost:3000/api/users`

- To view a profile, navigate to `localhost:3000/api/profiles/:username`, 

- To Edit a profile, navigate to `localhost:3000/api/profiles/:username`

#### Any background context you want to provide?

- N/A

#### What are the relevant pivotal tracker stories?
[#166789996](https://www.pivotaltracker.com/story/show/166789996)
#### Screenshots (if appropriate)

- View user profile
Screen Shot 2019-07-17 at 12 30 45

- Edit user profile
Screen Shot 2019-07-17 at 12 31 02

#### Checklist

- [x] My code follows the style guidelines of this project
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I don't have more than 3 major commits on this PR

### CODE REVIEW

File: controllers/profile.js [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+          error: 'User does not exists!',
+        });
+      }
+      const profile = await user.getProfile();
+      return res.status(200).json({
+        message: 'User profile retrieved!',
+        profile
+      });
+    } catch (error) {
+      next(error);
```
Comment (kodek-sleuth) [inline controllers/profile.js]: Sometimes this error is a very verbose sequelize error, its best to capture all error codes sent on this block and send a custom error message

**Review by kodek-sleuth — CHANGES_REQUESTED** [review]

File: controllers/profile.js [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+      const { username } = req.params;
+      const {
+        firstname, lastname, bio, image, phone, facebook, twitter, linkedIn, instagram, location
+      } = req.body;
+
+      const user = await users.findOne({
+        where: { username }
+      });
+
+      if (!user) {
```
Comment (kodek-sleuth) [inline controllers/profile.js:77]: Use if user === null || user === undefined

**Review by kodek-sleuth — CHANGES_REQUESTED** [review]

File: routes/api/index.js [hunk 3]
```diff
@@ -1,10 +1,12 @@
 import express from 'express';
 import admin from './admin';
 import users from './users';
+import profile from './profile';
 
 const router = express.Router();
 
 router.use('/', admin);
 router.use('/', users);
+router.use('/', profile);
```
Comment (kodek-sleuth) [inline routes/api/index.js]: Profile is in the Users class and should therefore be in the users.js file

**Review by kodek-sleuth — CHANGES_REQUESTED** [review]

**Review by elemanhillary-zz — CHANGES_REQUESTED** [review]
increase on coverage

File: controllers/profile.js [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+        facebook,
+        twitter,
+        linkedIn,
+        instagram,
+        location
+      });
+
+      return res.status(200).json({
+        message: 'User profile updated!',
+        updatedProfile
```
Comment (kodek-sleuth) [inline controllers/profile.js]: place this in a user object like user: {
  updatedProfile
}

**Review by kodek-sleuth — DISMISSED** [review]

File: package.json [hunk 5]
```diff
@@ -16,6 +16,12 @@
     "lint": "eslint . --ext .js",
     "migrate:undo": "node_modules/.bin/sequelize db:migrate:undo:all"
   },
+  "nyc": {
+    "exclude": [
+      "middlewares/validations/validations.js",
```
Comment (kodek-sleuth) [inline package.json]: please don't exclude the validations

**Review by kodek-sleuth — CHANGES_REQUESTED** [review]

**Review by elemanhillary-zz — DISMISSED** [review]
clean, ready for merge

**Review by elemanhillary-zz — CHANGES_REQUESTED** [review]
resolve conflicts

File: .codeclimate.yml [hunk 6]
```diff
@@ -0,0 +1,26 @@
+version: "2"
+exclude_patterns:
+- "node_modules"
+- "middlewares/validations/validations.js"
+- "controllers/profile.js"
```
Comment (elemanhillary-zz) [inline .codeclimate.yml]: kindly, why are you excluding your controller and middlewares to bypass `codeclimate`

**Review by elemanhillary-zz — CHANGES_REQUESTED** [review]

**Review by niyoceles — DISMISSED** [review]

**Review by niyoceles — DISMISSED** [review]

**Review by kodek-sleuth — DISMISSED** [review]
this pr is ready to be merged

**Review by elemanhillary-zz — DISMISSED** [review]
ready for merge

**Review by kodek-sleuth — DISMISSED** [review]
This Pr is ready to be merged

**Review by niyoceles — DISMISSED** [review]

File: controllers/profile.js [hunk 7]
```diff
@@ -0,0 +1,116 @@
...
+  * @static
+  */
+  static async getProfile(req, res) {
+    try {
+      const { username } = req.params;
+      const user = await users.findOne({
+        where: { username }
+      });
+
+      if (user === null || user === undefined) {
```
Comment (elemanhillary-zz) [inline controllers/profile.js]: replace this with `!user` instead having two operands that are similar in logic

**Review by elemanhillary-zz — CHANGES_REQUESTED** [review]
replace it everywhere neccessary

**Review by elemanhillary-zz — DISMISSED** [review]
ready for merge

**Review by Hervera — DISMISSED** [review]
Ready to be marged

**Review by niyoceles — DISMISSED** [review]

**Review by Hervera — COMMENTED** [review]
Ready to be merged

**Review by niyoceles — DISMISSED** [review]

**Review by kodek-sleuth — DISMISSED** [review]

**Review by elemanhillary-zz — DISMISSED** [review]
clean, ready for merge

File: controllers/profile.js (hunk 7 above)
Comment (IsaiahRn) [inline controllers/profile.js]: Implement Hillary's feedback

File: controllers/profile.js [hunk 8]
```diff
@@ -0,0 +1,116 @@
...
+      const { email } = req.decoded;
+      const {
+        firstName, lastName, bio, phone, facebook, twitter, linkedIn, instagram, location
+      } = req.body;
+
+      const user = await users.findOne({
+        where: { email }
+      });
+
+      if (user === null || user === undefined) {
```
Comment (IsaiahRn) [inline controllers/profile.js]: Implement Hillary's feedback in line 26

File: controllers/profile.js [hunk 9]
```diff
@@ -0,0 +1,116 @@
...
+        return res.status(404).json({
+          error: 'User does not exists!',
+        });
+      }
+
+      await user.update({
+        firstName,
+        lastName,
+        bio,
+        image: (req.file ? req.file.url : req.user.image),
```
Comment (IsaiahRn) [inline controllers/profile.js]: change `req.user.image` to `user.image`

File: middlewares/imageUpload.js [hunk 10]
```diff
@@ -10,7 +10,7 @@ cloudinary.config({
 });
 const storage = cloudinaryStorage({
   cloudinary,
-  folder: 'articles',
+  folder: 'uploads',
   allowedFormats: ['jpg', 'png'],
   transformation: [{ width: 500, height: 500, crop: 'limit' }]
```
Comment (IsaiahRn) [inline middlewares/imageUpload.js]: remove transformation

File: middlewares/validations/authValidations.js [hunk 11]
```diff
@@ -73,6 +77,27 @@ class authValidations {
     next();
   }
 
+  static async validateProfile(req, res, next) {
+    const {
+      firstName, lastName
+    } = req.body;
+    const name = /^[a-zA-Z]*$/;
```
Comment (IsaiahRn) [inline middlewares/validations/authValidations.js]: change `name` to `nameRegex`

**Review by niyoceles — DISMISSED** [review]

**Review by Hervera — DISMISSED** [review]
Ready to be merged

File: controllers/profile.js [hunk 12]
```diff
@@ -0,0 +1,116 @@
...
+
+      if (!user) {
+        return res.status(404).json({
+          error: 'User does not exists!',
+        });
+      }
+
+      return res.status(200).json({
+        message: 'User profile retrieved!',
+        profile: {
```
Comment (dmithamo) [inline controllers/profile.js:38]: Destructure the `user` object for all these params

File: controllers/profile.js [hunk 13]
```diff
@@ -0,0 +1,116 @@
...
+        twitter,
+        linkedIn,
+        instagram,
+        location
+      });
+
+
+      return res.status(200).json({
+        message: 'User profile updated!',
+        profile: {
```
Comment (dmithamo) [inline controllers/profile.js:99]: See comment on line `34`

**Review by dmithamo — CHANGES_REQUESTED** [review]

**Review by niyoceles — APPROVED** [review]
It can be merged.

**Review by Hervera — APPROVED** [review]
Ready to be marged

**Review by elemanhillary-zz — APPROVED** [review]
ready for merge

**Review by dmithamo — APPROVED** [review]

**Pull request merged**